### ISSUE TITLE
user-defined RPC cannot return a decimal64 value

### ISSUE THREAD

**Issue by jktjkt** [CESNET/netopeer2#543]
I'm trying to execute a user-defined RPC over NETCONF with the following [YANG model: `czechlight-calibration-device.yang`](https://github.com/CESNET/Netopeer2/files/4130542/czechlight-calibration-device.yang.txt). The important part of that model is a user-defined RPC that returns one `decimal64` value.

When my sysrepo callback returns one leaf, `power`, passing a `double` value back to sysrepo, then `netopeer2-server` hits the following error:

```
netopeer2-server[202]: Schema node not found. (/czechlight-calibration-device:measure/power)
netopeer2-server[202]: Invalid value "" in "power" element. (/czechlight-calibration-device:measure/power)
netopeer2-server[202]: Failed to create node "power" as a child of "measure". (/czechlight-calibration-device:measure/power)
```
Over NETCONF, this is what I send:
```
<rpc xmlns="urn:ietf:params:xml:ns:netconf:base:1.0" message-id="35">
  <measure xmlns="http://czechlight.cesnet.cz/yang/calibration-device">
    <port>36</port>
  </measure>
</rpc>
```
And this is what I get back:
```xml
<rpc-reply xmlns="urn:ietf:params:xml:ns:netconf:base:1.0" message-id="35">
  <rpc-error>
    <error-type>protocol</error-type>
    <error-tag>bad-element</error-tag>
    <error-severity>error</error-severity>
    <error-message xml:lang="en">Failed to create node "power" as a child of "measure".</error-message>
    <error-info>
      <bad-element>/czechlight-calibration-device:measure/power</bad-element>\
    </error-info>
  </rpc-error>
</rpc-reply>
```
I *think* I understand how RPCs in sysrepo work, we have other stuff which works correctly. Also, I can trigger this RPC directly via the sysrepo APIs, and I can retrieve a correct result:
```c++
    auto conn = std::make_shared<sysrepo::Connection>("rpc");
    auto session = std::make_shared<sysrepo::Session>(conn);

    auto rpc = "/czechlight-calibration-device:measure"s;
    auto input = std::make_shared<sysrepo::Vals>(1);
    input->val(0)->set((rpc + "/port").c_str(), static_cast<uint8_t>(port));
    auto res = session->rpc_send(rpc.c_str(), input);
    assert(res->val_cnt() == 1);
    auto result = res->val(0)->data()->get_decimal64();
    std::cout << result << std::endl;
```

In other words, it's just the `netopeer2-server` that complains about the RPC result. Am I doing something wrong, perhaps?

This is with CESNET/libyang@b3a4d48eb8ff9d49c212a3c0922935d838398c71, sysrepo/sysrepo@7960e5d3925fbeb180fffd77562e8f1817949c44 and Netopeer2 f213cadad2ad362f41584168a82c513af8187c5b

**Comment by jktjkt** [comment]
Hmmm, by trial and error I found out that returning `uint32` instead of `decimal64` makes it work (if I change both the YANG model *and* the type of data I pass via sysrepo, of course).

**Comment by jktjkt** [comment]
I think I found the problem, but I would prefer that you fix it. Here's what happens:

- in [`op_generic.c`](https://github.com/CESNET/Netopeer2/blob/legacy/server/op_generic.c#L40), before `lyd_new_path` gets called, there's a call to `op_get_srval`
- `op_get_srval` does not know that we're handling RPC output anymore, so it [calls `ly_ctx_get_node` unconditionally referring to RPC's input, not output](https://github.com/CESNET/Netopeer2/blob/f213cadad2ad362f41584168a82c513af8187c5b/server/operations.c#L1874)

A *very* crude fix (e.g., it would break if there is an input and output parameter with the same name, but a different type):
```patch
diff --git a/server/operations.c b/server/operations.c
index 25fd9ed..8766aad 100755
--- a/server/operations.c
+++ b/server/operations.c
@@ -1873,7 +1873,11 @@ op_get_srval(struct ly_ctx *ctx, const sr_val_t *value, char *buf)
         /* get fraction-digits */
         sleaf = (struct lys_node_leaf *)ly_ctx_get_node(ctx, NULL, value->xpath, 0);
         if (!sleaf) {
-            return NULL;
+            sleaf = (struct lys_node_leaf *)ly_ctx_get_node(ctx, NULL, value->xpath, 1);
+            if (!sleaf) {
+                return NULL;
+            }
+            ly_errno = LY_SUCCESS;
         }
         while (sleaf->type.base == LY_TYPE_LEAFREF) {
             sleaf = sleaf->type.info.lref.target;
```
I think that a proper fix is to extend that function with some flags for distinguishing between input and output.

There's one more issue which is libyang's C++ bindings' lack of an accessor to the actual `decimal64` value when one only has a `libyang::S_Value` (I don't have access to the number of decimal digits). I'll submit a patch or a bug about that in libyang.

**Comment by michalvasko** [comment]
Hi,
you are right and it should be fixed now the way you proposed but I have not tested it. As for the `decimal64` in `S_Value`, you are also right that it is missing. I guess the reason is the same as why `bit` was missing (you need `type` to get the value) except it was not even commented out so I have not added it when I added `bit`. So okay, please provide a pull request. Even though you should not need to change anything in the C++ bindings (only add the method), it may not be that simple. Take a look [here](https://github.com/CESNET/libyang/blob/devel/src/parser_lyb.c#L644) (but do not forget about [leafrefs](https://github.com/CESNET/libyang/blob/devel/src/parser_lyb.c#L550)).

Regards,
Michal

**Comment by jktjkt** [comment]
@syyyr, do you feel like adding support for `decimal64` -> `double` into libyang's C++ bindings?

**Comment by syyyr** [comment]
Oof

**Comment by michalvasko** [comment]
@jktjkt You yourself have proposed to fix it but it is no problem, I can add the function (but please open an issue). Now that I think about it, the function can be added but there is a good chance that by transforming `decimal64` into `double` you lose some precision.

Regards,
Michal
